### PR TITLE
hcq graph

### PR DESCRIPTION
**Pull request by nimlgen**
Started to experiment with HCQ graph for NV and AMD to understand what API we want to unify them

### CODE REVIEW

File: tinygrad/runtime/graph/hcq.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import ctypes, collections, time, itertools
+from typing import List, Any, Dict, cast, Optional, Union, Tuple
+from tinygrad.helpers import GraphException, init_c_var, round_up, colored
+from tinygrad.buffer import Buffer, BufferOptions
+from tinygrad.device import Compiled, CompiledRunner, BufferXfer, MultiDeviceJITGraph, Device, Runner
+from tinygrad.shape.symbolic import Variable
+from tinygrad.engine.realize import ExecItem
+from tinygrad.engine.jit import get_input_replace, get_jit_stats, get_jc_idxs_with_updatable_launch_dims, get_jc_idxs_with_updatable_var_vals
+
+class HCQGraph(Runner):
```
Comment (geohot) [inline tinygrad/runtime/graph/hcq.py:10]: Check out the new GraphRunner class

**Comment by geohot** [comment]
How close is this to working? Is it far enough along where you could try ResNet? Ideally I want the MLPerf submissions to use this.

**Comment by nimlgen** [comment]
Quite close already hlb_cifar seems to work fine (will push it). Going to try NV on resnet today. 

Still thinking about the final design synchronization to use for both NV and AMD: use binary semaphores (mostly like HSA) or "timeline" semaphores (where wait passes when `>=`). I think timeline semaphores are better in some ways (like no need to reset on CPUs, less signals to allocate), but it has some logic cost to reset it to 0 at the beginning of each graph run (on the GPU) if I am not missing something obvious and simple solution (will think more on this, how to make it simpler).

**Comment by geohot** [comment]
Oh that's a neat idea, I didn't consider timeline semaphores. I think that's preferred since it's O(1) (I'm fine with O(1) reset). Don't you also have to reset binary semaphores?

**Comment by nimlgen** [comment]
First resnet runs with this. I think it's possible to optimize it more:
AMD https://wandb.ai/nimlgen/tinygrad-examples_mlperf/runs/am7bavet
NV https://wandb.ai/nimlgen/tinygrad-examples_mlperf/runs/w0l52i8h

**Comment by nimlgen** [comment]
Things to do in separate prs:
1) NV sync speed up (sync using qmd). Tested already this makes NV's gpu speed the same as CUDA's.
2) Optimize how we deal with updatable launch dims. Now we build a sep hcq for each of them: we can patch hcq for amd and qmd for nv, should be way faster python time.
3) Amd sdma fast submit() + some python optimization.

**Comment by nimlgen** [comment]
@geohot, want to review? I keep this as minimal hcq only PR, will do speed related changes in the next prs.

**Pull request merged**